src/vm/database/marf.rs

```diff
@@ -709,3 +660,233 @@ impl ClarityBackingStore for NullBackingStore {
         panic!("NullBackingStore cannot put")
     }
 }
+
+impl<'a> WritableMarfStore<'a> {
+    pub fn as_clarity_db<'b>(
+        &'b mut self,
+        headers_db: &'b dyn HeadersDB,
+        burn_state_db: &'b dyn BurnStateDB,
+    ) -> ClarityDatabase<'b> {
+        ClarityDatabase::new(self, headers_db, burn_state_db)
+    }
+
+    pub fn as_analysis_db<'b>(&'b mut self) -> AnalysisDatabase<'b> {
+        AnalysisDatabase::new(self)
+    }
+
+    pub fn rollback_block(self) {
+        self.marf.drop_current();
+    }
+
+    pub fn rollback_unconfirmed(self) {
+        SqliteConnection::drop_metadata(self.marf.sqlite_tx(), &self.chain_tip);
+        self.marf.drop_unconfirmed();
```

Comment:
This all happens within the context of a single DB transaction, right?

Reply:
Yep -- per my previous comment, the `MarfTransaction` instance held by this struct represents a single DB transaction, so both `drop_metadata` and `drop_unconfirmed` happen in the same transaction (as do all the prior operations applied during the lifetime of the struct).

For unconfirmed tries (unlike the anchored block tries), this can leave the side store _somewhat_ dirty, because unconfirmed tries need to commit (the anchored block tries just abort the transaction: the entirety of their writes exist in one sqlite tx) -- there would be left-over side store data, but not any metadata (because it's dropped here). The left-over side store data is safe, because that data is just mapping `hash(value) -> value` (though it is eating up storage space), but adding garbage collection there can be handled later (it won't affect consensus).

Reply:
Thanks for explaining!  As long as there aren't any rows in the side-store that could conflict with subsequent unconfirmed (or confirmed) state, then this is fine.  Can you open an issue to add garbage-collection for unconfirmed side-store state?